storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -671,8 +672,9 @@ private LogOffsetMetadata fetchLastStableOffsetMetadata() throws IOException {
      * beyond the high watermark.
      */
     public long lastStableOffset() {
-        if (firstUnstableOffsetMetadata.isPresent() && firstUnstableOffsetMetadata.get().messageOffset < highWatermark()) {
-            return firstUnstableOffsetMetadata.get().messageOffset;
+        Optional<LogOffsetMetadata> firstUnstableOffsetMetadataCopy = firstUnstableOffsetMetadata;
```

Comment:
Maybe worth adding a comment. There are a few places in this class where we say we cache a value to avoid concurrency issues.

Reply:
Done